### DISCUSSION TITLE
Fastest way to upload, reset and follow output

### DISCUSSION THREAD

**Discussion by acidkewpie** [micropython/micropython#13190] [Using MicroPython] [answered]
Hi, I've been playing around with all sorts of tools, mpr, ampy, rshell, mpremote but for some reason I don't understand I can't work out a simple work flow to...

1) upload a new main.py
2) soft-reset
3) connect to repl to see what it's up to

I can then easily loop this and kill the repl session with inotifywatch or such like when I save my main.py on my laptop

Quirks I'm seeing, which I expect are obvious...

Connecting after a soft-rest, repl shows nothing, even though I thought mpremote resume" was meant to pick up on that existing interpreter?

If I do a control D to do a soft reset inside repl, then it starts up as expected, but I don't want to have to do every time.

If I do a full reset I can then connect to repl and see what it's up to, but this takes time, and destroys the ttyACM0 device during that reboot.

when using "run" commands in various ways, my code terminates immediately, rather than waiting for the timers I have just configured. I have no infinite loop as I don't need one when the code runs normally outside of a dev environment.

**Comment by andrewleech** [comment]
The quickest development workflow I use is to have my code in a different file, eg `app.py` and not have it load automatically from main.

Then use mpremote to mount my local folder with `app.py` on the device and run it manually from repl.

Make changes in local file, ctrl-d to soft reset and reload changes on mount, then run it again.

Eg.
```
mpremote mount . repl --inject-code "import app\n"
```
After mounting the current local directory, it executes app.py from the mounted directory each time Ctrl-J is pressed.

**Comment by acidkewpie** [comment]
I was keen to keep the file on board, so it will run the same when i disconnect it, and also to be able to run things totally hands free on that side. 

I've got this nonsense running currently, just want to turn the reset into a soft-reset really. 

```
( while [ -c /dev/ttyACM0 ]
  do inotifywait main.py -qq -e DELETE_SELF
    echo "main.py changed, reloading"
   fuser -sk /dev/ttyACM0 2>/dev/null
   sleep 10
 done ) & 

while true
  do mpremote cp main.py :main.py + reset
  inotifywait -e close -qq --include ttyACM0 /dev/
  echo "device recreated, starting mpremote again"
  mpremote
done
```

I probably just shouldn't resent a couple of extra keypresses I guess! I wasn't aware of the inject thing, i'll have a play around with that,

Am I right to observe that your import approach won't work with timers though? it just gets to the end of the file and stops. There's something different in importing a file in repl like that compared to executing main.py on boot isn't there?

**Comment by andrewleech** [comment]
The mpremote mount stuff is just for during active development sure, once I've got the code right I'll mpremote cp to board and add a `import app` into my main.py to run it automatically.

code imported in a file like `app.py` in repl or main runs exactly the same as code written directly into main.
I use timers etc no worries.
If I want long running code in app I just have an infinite while loop or `asyncio.run(main())` etc.

**Comment by jimmo** [comment]
>  I thought `mpremote resume` was meant to pick up on that existing interpreter

`resume` is implicit for the `repl` command (which is the default command).

https://docs.micropython.org/en/latest/reference/mpremote.html#auto-connection-and-soft-reset

If you've just done a file copy, then the current state will be sitting at the REPL. Which is why you then need to press Ctrl-D to start your program.

(Note the soft-reset triggered by mpremote is special in that it doesn't cause main.py to be executed)

It sounds like what you want is an explicit "run main.py" command (or a version of the `soft-reset` command that does run main.py).

**Comment by bulletmark** [comment, marked as answer]
> (Note the soft-reset triggered by mpremote is special in that it doesn't cause main.py to be executed)

I've always been confused what/if special magic is done here. This really should be stated in the mpremote documentation, within that [section](https://docs.micropython.org/en/latest/reference/mpremote.html#auto-connection-and-soft-reset).

**Reply by jimmo** [reply]
`mpremote` (and pyboard.py) operate in the "raw" REPL mode when issuing commands etc. In raw mode, on soft reset, main.py isn't run.

(This is another reason why as little code as possible should be in boot.py, because it will run on every mpremote command)

**Reply by acidkewpie** [reply]
Every mpremote command? I migrated a wifi captive portal into my boot.py and that definitely doesn't run every time. Was there a more specific version of "every" here?

**Reply by jimmo** [reply]
@acidkewpie See the link above https://docs.micropython.org/en/latest/reference/mpremote.html#auto-connection-and-soft-reset
